### PR TITLE
[features] Add PopSIFT ImageDescriber

### PR DESCRIPTION
**Pull request by gregoire-dl**
- rename SIFT_describer into SIFT_vlfeat_describer
- rename SIFT_float_describer into SIFT_vlfeat_float_describer
- add SIFT_popSIFT_describer
- add generic SIFT_describer which use SIFT_vlfeat_describer or
SIFT_popSIFT_describer if the user uses CUDA
- move SIFT imageDescribers into openMVG/features/sift
- remove cereal from imageDescribers (was no more used)

### CODE REVIEW

File: src/openMVG/features/sift/SIFT_popSIFT_describer.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include <openMVG/logger.hpp>
+
+namespace openMVG {
+namespace features {
+
+bool SIFT_popSIFT_ImageDescriber::Describe(const image::Image<unsigned char>& image,
+                                      std::unique_ptr<Regions>& regions,
+                                      const image::Image<unsigned char>* mask)
+{
+  popsift::cuda::device_prop_t deviceInfo;
```
Comment (gregoire-dl) [inline src/openMVG/features/sift/SIFT_popSIFT_describer.cpp]: @griwodz 
In the RT camera tracker, we first create all imageDescribers and we call the `Describe` method in a dedicated thread per camera. We have one thread per camera, but these threads are generated by tbb at each frame.
So we put the call to `cudaSetDevice(_cudaPipeIndex)` (which is done inside `popsift::cuda::device_prop_t::set`) in the `Describe` function because cudaSetDevice relies on the calling host thread.

Currently the code works fine with one camera. But if we use multiple cameras, we get : 

```
Choosing device Choosing device 00: GeForce GTX 980 Ti: GeForce GTX 980 Ti

Device information:Device information:
    Name: GeForce GTX 980 Ti
    Compute Capability:    5
    Name: GeForce GTX 980 Ti
    Compute Capability:    5.2
    Total device mem:      .2
    Total device mem:      6372589568 B 6223232 kB 6077 MB 
    Per-block shared mem:  49152
    Warp size:             32
    Max threads per block: 1024
    Max threads per SM(X): 2048
    Max block sizes:       {1024,1024,64}
    Max grid sizes:        {2147483647,65535,65535}
    Number of SM(x)s:      22
    Concurrent kernels:    yes
    Mapping host memory:   yes
    Unified addressing:    yes

6372589568 B 6223232 kB 6077 MB 
    Per-block shared mem:  49152
    Warp size:             32
    Max threads per block: 1024
    Max threads per SM(X): 2048
    Max block sizes:       {1024,1024,64}
    Max grid sizes:        {2147483647,65535,65535}
    Number of SM(x)s:      22
    Concurrent kernels:    yes
    Mapping host memory:   yes
    Unified addressing:    yes

/s/apps/packages/mikrosVfx/multiview/popsift/develop/repo/src/popsift/sift_desc.cu:/s/apps/packages/mikrosVfx/multiview/popsift/develop/repo/src/popsift/sift_extremum.cu:8370 Runtime error:
    Failed to (re)allocate memory for downloading -1815989562 descriptors

    cudaEventRecord failed: an illegal memory access was encountered
    Not enough memory.
```

Do you have an idea how we should use it?

File: src/openMVG/features/akaze/AKAZE.hpp [hunk 2]
```diff
@@ -55,22 +55,11 @@ struct AKAZEConfig
   {
   }
 
-  template<class Archive>
-  void serialize(Archive & ar)
```
Comment (pmoulon) [inline src/openMVG/features/akaze/AKAZE.hpp]: Why removing this?

File: src/openMVG/features/sift/SIFT_vlfeat_describer.hpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
+#pragma once
```
Comment (pmoulon) [inline src/openMVG/features/sift/SIFT_vlfeat_describer.hpp]: License! you cannot delete the file in nonfree and copy paste here and removing the license!

File: src/openMVG/features/akaze/AKAZE.hpp (hunk 2 above)
Comment (fabiencastan) [inline src/openMVG/features/akaze/AKAZE.hpp]: We don't use image describers serialization anymore in the pipeline.

File: src/openMVG/features/sift/SIFT_vlfeat_describer.hpp (hunk 3 above)
Comment (fabiencastan) [inline src/openMVG/features/sift/SIFT_vlfeat_describer.hpp]: As discussed before, all files are still protected by the license of the project. But we will run a script soon to update them and clarify the files that are part of the project.

File: src/openMVG/features/sift/SIFT_describer.hpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+namespace openMVG {
+namespace features {
+
+class SIFT_ImageDescriber : public Image_describer
+{
+public:
+  SIFT_ImageDescriber(const SiftParams& params = SiftParams(), bool isOriented = true)
+    : _params(params)
+  {
+    // TODO: detect if CUDA is available on the computer
```
Comment (fabiencastan) [inline src/openMVG/features/sift/SIFT_describer.hpp]: @griwodz: Is there a function in popSift or directly in cuda for that?

File: src/openMVG/features/akaze/AKAZE.hpp (hunk 2 above)
Comment (pmoulon) [inline src/openMVG/features/akaze/AKAZE.hpp]: How do you track which method have been used for a scene?

File: src/openMVG/features/sift/SIFT_describer.hpp (hunk 4 above)
Comment (pmoulon) [inline src/openMVG/features/sift/SIFT_describer.hpp]: You certainly have at least a compilation guard (preprocessor)

File: src/openMVG/features/akaze/AKAZE.hpp (hunk 2 above)
Comment (fabiencastan) [inline src/openMVG/features/akaze/AKAZE.hpp]: We have added the support to use multiple describers in the reconstruction. So we keep the describer type information during the matching, but also in Tracks and Landmarks.
But we don't keep the information about the parameters used for each method.

File: src/openMVG/features/sift/SIFT_describer.hpp (hunk 4 above)
Comment (simogasp) [inline src/openMVG/features/sift/SIFT_describer.hpp]: popsift can only build with CUDA, so if it is found it means that cuda is available

Comment (fabiencastan) [inline src/openMVG/features/sift/SIFT_describer.hpp]: If it's builded without CUDA, it's always false but that doesn't help.
If it's builded with CUDA enabled, it doesn't mean that it will run on an hardware supporting CUDA.
The goal is to build it once and use it on multiple hardwares.

Comment (pmoulon) [inline src/openMVG/features/sift/SIFT_describer.hpp]: What about considering to build a similar interface to https://github.com/openMVG/openMVG/blob/develop/src/openMVG/system/cpu_instruction_set.hpp to check GPU capabilities?

Comment (fabiencastan) [inline src/openMVG/features/sift/SIFT_describer.hpp]: Yes, something like that: https://github.com/alicevision/popsift/blob/develop/src/popsift/common/device_prop.cpp

Comment (fabiencastan) [inline src/openMVG/features/sift/SIFT_describer.hpp]: @gregoire-dl: Could you make another PR with this test?
Just need to list the cuda devices and if there no one, we have to use the cpu version.
```
    int numDevices;
    cudaError_t err = cudaGetDeviceCount(&numDevices);
```
